test/connectivity_state_updates_test.go

```diff
@@ -0,0 +1,121 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package test
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/internal"
+	"google.golang.org/grpc/internal/grpcsync"
+	"google.golang.org/grpc/internal/stubserver"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/resolver/manual"
+)
+
+type testSubscriber struct {
+	onMsgCh chan connectivity.State
+}
+
+func newTestSubscriber() *testSubscriber {
+	return &testSubscriber{onMsgCh: make(chan connectivity.State, 1)}
+}
+
+func (ts *testSubscriber) OnMessage(msg interface{}) {
+	select {
+	case ts.onMsgCh <- msg.(connectivity.State):
+	default:
+	}
+}
+
+func (s) TestConnectivityStateUpdates(t *testing.T) {
+	// Create a ClientConn with a short idle_timeout.
+	r := manual.NewBuilderWithScheme("whatever")
+	dopts := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithResolvers(r),
+		grpc.WithIdleTimeout(defaultTestShortIdleTimeout),
+		grpc.WithDefaultServiceConfig(`{"loadBalancingConfig": [{"round_robin":{}}]}`),
+	}
+	cc, err := grpc.Dial(r.Scheme()+":///test.server", dopts...)
+	if err != nil {
+		t.Fatalf("grpc.Dial() failed: %v", err)
+	}
+	t.Cleanup(func() { cc.Close() })
+
+	// Start a test backend and push an address update via the resolver.
+	backend := stubserver.StartTestService(t, nil)
+	t.Cleanup(backend.Stop)
+
+	s := newTestSubscriber()
+	internal.SubscribeToConnectivityStateChanges.(func(cc *grpc.ClientConn, s grpcsync.Subscriber) func())(cc, s)
+
+	expectedStates := []connectivity.State{
+		connectivity.Connecting,
+		connectivity.Ready,
+		connectivity.Idle,
+		connectivity.Shutdown,
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() {
+		for _, expectedState := range expectedStates {
+			select {
+			case actualState := <-s.onMsgCh:
+				if actualState != expectedState {
+					t.Errorf("Received unexpected state: %q; want: %q", actualState, expectedState)
+				}
+			case <-time.After(defaultTestTimeout):
+				t.Error("Timeout when expecting the onMessage() callback to be invoked")
+			}
+			if t.Failed() {
+				break
+			}
+		}
+		wg.Done()
+	}()
+
+	go func() {
+		r.UpdateState(resolver.State{Addresses: []resolver.Address{{Addr: backend.Address}}})
+
+		// Verify that the ClientConn moves to READY.
+		ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+		defer cancel()
+		awaitState(ctx, t, cc, connectivity.Ready)
+
+		// Verify that the ClientConn moves to IDLE as there is no activity.
+		awaitState(ctx, t, cc, connectivity.Idle)
+
+		cc.Close()
+		awaitState(ctx, t, cc, connectivity.Shutdown)
+
+		wg.Done()
+	}()
+
+	wg.Wait()
+	if t.Failed() {
+		t.FailNow()
+	}
```

Comment:
Can this be done instead in the check for `t.Failed` in the first goroutine. Because I don't see any call to `t.Error` in the second one. So, it might just be easier to fail from the first goroutine when an unexpected state is received.

Reply:
If so, the first goroutine will only fail, not the main goroutine.
We need to fail after `wg.Wait()` in order to wait finishing the first goroutine.